src/nvim/buffer.c

```diff
@@ -357,6 +357,7 @@ void set_bufref(bufref_T *bufref, buf_T *buf)
 ///
 /// @param bufref Buffer reference to check for.
 bool bufref_valid(bufref_T *bufref)
+  FUNC_ATTR_PURE
```

Comment:
https://github.com/neovim/nvimdev.nvim should probably highlight this in orange or something. I always wonder if someone might accidentally change one of these functions to cause side effects, without removing the attr.

Reply:
Probably a matter of time. Maybe we should make some benchmarks to see if the benefits are worth it.